File: trees/serialize_and_deserialize_binary_tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:
    def serialize(self, root):
        self.pre_order = []
        self.dfs_serialize(root)
        return ",".join(self.pre_order)

    def dfs_serialize(self, node):
        if not node: return self.pre_order.append('#')

        self.pre_order.append(str(node.val))
        self.dfs_serialize(node.left)
        self.dfs_serialize(node.right)

    def deserialize(self, data):
        if not data: return None

        pre_order = data.split(',')
        return self.dfs_deserialize(pre_order)

    def dfs_deserialize(self, pre_order):
        if not pre_order: return None

        if pre_order[0] == '#':
            pre_order.pop(0)
            return None

        root = TreeNode(int(pre_order.pop(0)))
        root.left = self.dfs_deserialize(pre_order)
        root.right = self.dfs_deserialize(pre_order)

        return root

# Approach 2: BFS pre-order / Iterative approach (Seems faster on LC inputs)
import collections
class Codec:
    def serialize(self, root):
        if not root: return []
        result = []
        que = collections.deque([root])
        while que:
            node = que.popleft()
            if not node:
                result.append('#')
                continue
            else:
                result.append(str(node.val))
            que.extend([node.left, node.right])
        return ' '.join(result)

    def deserialize(self, data):
        if not data: return None
        iterData = iter(data.split(' '))
        root = TreeNode(next(iterData))
        que = collections.deque([root])
        while que:
            node = que.popleft()
            if not node: continue
            val = next(iterData)
            node.left = TreeNode(int(val)) if val != '#' else None
            que.append(node.left)
            val = next(iterData)
            node.right = TreeNode(int(val)) if val != '#' else None
            que.append(node.right)
        return root


# 297. Serialize and Deserialize Binary Tree
# https://leetcode.com/problems/serialize-and-deserialize-bst/description/

# Approach 1: DFS using pre-order
# Steps:
# 1. Serialize using pre-order traversal using depth first search and represent null by "#"
# 2. Deserialize by traversing the serialized tree and using recursion. O(n).

# Approach 2: BFS using queue and iterative approach and pre_order
# Time: O(n), visit n nodes once
# Space: O(n), store the entire tree in array and string

#     1
#    / \
#   2   3
#      / \
#     4   5

root = TreeNode(1)
l, r = TreeNode(2), TreeNode(3)
r1, r2 = TreeNode(4), TreeNode(5)
root.left, root.right = l, r
r.left, r.right = r1, r2

codec = Codec()
assert codec.serialize(root) == "1,2,#,#,3,4,#,#,5,#,#"
new_root = codec.deserialize("1,2,#,#,3,4,#,#,5,#,#")

assert new_root.val == root.val
assert new_root.left.val == l.val
assert new_root.right.val == r.val
assert new_root.right.left.val == r1.val
assert new_root.right.right.val == r2.val

File: trees/serialize_and_deserialize_binary_tree.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# Encodes a tree to a single string.
# @param {TreeNode} root
# @return {string}
def serialize(root)
    pre_order = []
    dfs_serialize(root, pre_order)
    pre_order.join(",")
end

def dfs_serialize(node, pre_order)
    if !node
        pre_order.push("#")
        return
    end

    pre_order.push(node.val.to_s)
    dfs_serialize(node.left, pre_order)
    dfs_serialize(node.right, pre_order)
end


# Decodes your encoded data to tree.
# @param {string} data
# @return {TreeNode}
def deserialize(data)
    return if data.empty?
    pre_order = data.split(",")
    dfs_deserialize(pre_order)
end

def dfs_deserialize(pre_order)
    return if pre_order.empty?

    val = pre_order.shift()
    val == '#' ? return : val = val.to_i
    root = TreeNode.new(val)

    root.left = dfs_deserialize(pre_order)
    root.right = dfs_deserialize(pre_order)

    root
end

# 297. Serialize and Deserialize Binary Tree
# https://leetcode.com/problems/serialize-and-deserialize-bst/description/

# Approach 1: DFS using pre-order for serialization and DFS and Queue for deserialization
# Steps:
# 1. Serialize using pre-order traversal using depth first search and represent null by "#"
# 2. Deserialize by traversing the serialized tree string and using recursion. O(n).

# Approach 2: BFS using queue and iterative approach and pre_order
# Time: O(n), visit n nodes once
# Space: O(n), store the entire tree in array and string


require 'test/unit'
extend Test::Unit::Assertions

root = TreeNode.new(1)
l, r = TreeNode.new(2), TreeNode.new(3)
r1, r2 = TreeNode.new(4), TreeNode.new(5)
root.left, root.right = l, r
r.left, r.right = r1, r2

assert_equal(serialize(root), "1,2,#,#,3,4,#,#,5,#,#")

new_root = deserialize("1,2,#,#,3,4,#,#,5,#,#")
assert_equal(new_root.val, root.val)
assert_equal(new_root.left.val, l.val)
assert_equal(new_root.right.val, r.val)
assert_equal(new_root.right.left.val, r1.val)
assert_equal(new_root.right.right.val, r2.val)